states/visited2.go

```diff
@@ -0,0 +1,64 @@
+package states
+
+import (
+	"context"
+	"sync"
+
+	"github.com/earthly/earthly/domain"
+	"github.com/earthly/earthly/util/platutil"
+	"github.com/earthly/earthly/variables"
+)
+
+// VisitedCollections is a collection of visited targets.
+type VisitedCollection2 struct {
+	mu      sync.Mutex
+	visited map[string]*SingleTarget // targetInputHash -> sts
+	// Same collection as above, but as a list, to make the ordering consistent.
+	visitedList []*SingleTarget
+}
+
+// NewVisitedCollection returns a collection of visited targets.
+func NewVisitedCollection2() *VisitedCollection2 {
+	return &VisitedCollection2{
+		visited: make(map[string]*SingleTarget),
+	}
+}
+
+// All returns all visited items.
+func (vc *VisitedCollection2) All() []*SingleTarget {
+	vc.mu.Lock()
+	defer vc.mu.Unlock()
+	return append([]*SingleTarget{}, vc.visitedList...)
+}
+
+// Add adds a target to the collection, if it hasn't yet been visited. The returned sts is
+// either the previously visited one or a brand new one.
+// This function blocks if there is a visited target that is still running.
+func (vc *VisitedCollection2) Add(ctx context.Context, target domain.Target, platr *platutil.Resolver, allowPrivileged bool, overridingVars *variables.Scope, parentDepSub chan string) (*SingleTarget, bool, error) {
+	// Constructing a new sts early to be able to compute its target input hash.
+	newSts, err := newSingleTarget(ctx, target, platr, allowPrivileged, overridingVars, nil)
+	if err != nil {
+		return nil, false, err
+	}
+	newKey, err := newSts.targetInput.Hash()
+	if err != nil {
+		return nil, false, err
+	}
+	vc.mu.Lock()
+	sts, found := vc.visited[newKey]
+	if found {
+		vc.mu.Unlock()
```

Comment:
Looks like `newSts.targetInput.Hash()` is not consistent with `newSts.targetInput.Equals()`, that is, we are getting cache hits (same hash), that are not considered equal. Not sure yet of the consequences but worth mentioning.

Changing to this test file
```
VERSION 0.7

FROM earthly/dind:alpine

parallel:
  BUILD +test-executor
  BUILD +test-executor --INDEX=2
  BUILD +test-executor --INDEX=3
  BUILD +test-executor --INDEX=4
  BUILD +test-executor --INDEX=5

test-executor:
  ARG INDEX=1
  RUN --no-cache for i in $(seq 1 3); do  \
        echo "### $INDEX"; \
        sleep 0.3; \
      done;
  IF true
    RUN true
  END
```
and adding here this code:
```
if !sts.targetInput.Equals(newSts.targetInput) {
			nb, _ := json.Marshal(newSts.targetInput)
			b, _ := json.Marshal(sts.targetInput)
			fmt.Println(fmt.Sprintf("%s %s\ncreated:%s\nfound:  %s", target.String(), strings.Join(overridingVars.Sorted(), " "), string(nb), string(b)))
		}
```
outputs:
```
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"2","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"2","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"3","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"3","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./tests/with-docker-parallel+test-executor DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS INDEX
created:{"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"4","defaultConstant":""}],"platform":"linux/amd64","allowPrivileged":true}
found:  {"targetCanonical":"github.com/earthly/earthly/tests/with-docker-parallel:vlad/issue#2377+test-executor","buildArgs":[{"name":"DOCKERHUB_MIRROR_AUTH_FROM_CLOUD_SECRETS","constantValue":"true","defaultConstant":""},{"name":"INDEX","constantValue":"4","defaultConstant":"1"}],"platform":"linux/amd64","allowPrivileged":true}
./t
```

Reply:
This should be ok. The only difference is the `DefaultValue` field, which we want to ignore anyway in the new implementation.

Reply:
I would rather make `hash` and `equals` implementations consistent. 
What makes more sense, making `hash` dealing with the defaultConstant or `equals` ignoring it?

Reply:
Let me check if it's safe to drop it from equals - we really need hash to ignore it as part of this.

Reply:
Ok, I made these consistent now.